### PR TITLE
Allow changing the bug report url for the ice hook

### PR DESCRIPTION
**Pull request by bjorn3**
cc https://github.com/bjorn3/rustc_codegen_cranelift/issues/1174

### CODE REVIEW

**Comment by bjorn3** [comment]
I opened this PR, so I probably shouldn't review it myself. :)

r? @jackh726

**Comment by jackh726** [comment]
🤦‍♀️ whoops

File: compiler/rustc_driver/src/lib.rs [hunk 1]
```diff
@@ -1151,23 +1151,26 @@ pub fn catch_with_exit_code(f: impl FnOnce() -> interface::Result<()>) -> i32 {
...
 ///
 /// When `install_ice_hook` is called, this function will be called as the panic
 /// hook.
 pub fn report_ice(info: &panic::PanicInfo<'_>, bug_report_url: &str) {
-    // Invoke the default handler, which prints the actual panic message and optionally a backtrace
-    (*DEFAULT_HOOK)(info);
-
-    // Separate the output with an empty line
-    eprintln!();
-
```
Comment (jackh726) [inline compiler/rustc_driver/src/lib.rs:1170]: Why did this need to be taken out? I guess maybe so the right `DEFAULT_HOOK` gets registered?

**Review by jackh726 — COMMENTED** [review]
Honestly, I'm not sure how I feel about this. I feel like most ICEs happen in the frontend, not backend. So *usually* reporting to `rust-lang/rust` is a better idea. So maybe instead it's better to just include some extra information in the ICE report?

I'm probably not the best person to review this, since I'm not really in the loop here.

File: compiler/rustc_driver/src/lib.rs (hunk 1 above)
Comment (bjorn3) [reply compiler/rustc_driver/src/lib.rs:1170]: `DEFAULT_HOOK` both gets the default hook and registers the new ice hook with the default bug report url. This means that if you register a custom ice hook that calls `report_ice`, the ice hook with the default bug report url will be registered again. It is also easy to get into a cycle where the default ice hook and the custom ice hook recursively call each other on panics.

**Comment by bjorn3** [comment]
> I feel like most ICEs happen in the frontend, not backend. So usually reporting to rust-lang/rust is a better idea.

While this is true for the mature cg_llvm backend, this is not true for cg_clif. I do still sometimes get an ice caused by cg_clif. If the ice is in the frontend, I can redirect the issue to rust-lang/rust. In addition when using cg_clif, it is reasonable for someone to assume that cg_clif caused the issue if it doesn't reproduce with cg_llvm.

**Comment by jackh726** [comment]
I don't feel comfortable approving this myself. Maybe if someone from @rust-lang/compiler wants to weigh in or an MCP, that might be better.

**Comment by nikomatsakis** [comment]
I think an MCP would be a good idea here, but I approve the general concept.

**Comment by bjorn3** [comment]
> I think an MCP would be a good idea here

Opened https://github.com/rust-lang/compiler-team/issues/436

**Comment by oli-obk** [comment]
cc @rust-lang/clippy we may want this, too

**Comment by flip1995** [comment]
Clippy already has this since 2019 :+1: https://github.com/rust-lang/rust/blob/257782579915963c9dbe7433102275743837b9a8/src/tools/clippy/src/driver.rs#L155-L159

**Comment by jackh726** [comment]
MCP is finished

@bors r+ rollup

File: compiler/rustc_driver/src/lib.rs [hunk 2]
```diff
@@ -1151,23 +1151,26 @@ pub fn catch_with_exit_code(f: impl FnOnce() -> interface::Result<()>) -> i32 {
...
+        panic::set_hook(Box::new(|info| {
+            // Invoke the default handler, which prints the actual panic message and optionally a backtrace
+            (*DEFAULT_HOOK)(info);
+
+            // Separate the output with an empty line
+            eprintln!();
+
+            // Print the ICE message
+            report_ice(info, BUG_REPORT_URL);
+        }));
```
Comment (eddyb) [inline compiler/rustc_driver/src/lib.rs:1163]: This makes `report_ice` useless for the situations without enough control over the driver to bypass the `SyncLazy::force(&DEFAULT_HOOK);`. Maybe there should be a `report_ice_including_default_panic_hook` function that provides the old behavior?

Though ideally the new behavior should probably use a new function name (or better, both should change names, so use sites have to pick).

Comment (bjorn3) [reply compiler/rustc_driver/src/lib.rs:1163]: You can set a panic hook more than once, so you can first get the default hook, then let whatever code forces `DEFAULT_HOOK` run and then set the new panic hook.

Comment (bjorn3) [reply compiler/rustc_driver/src/lib.rs:1163]: A `report_ice_including_default_panic_hook` function requires the `DEFAULT_HOOK` to be forced first to be able to call it without ending in an infinite loop.

Comment (eddyb) [reply compiler/rustc_driver/src/lib.rs:1163]: > so you can first get the default hook

The codegen backend is loaded *after* `DEFAULT_HOOK` is initially forced, so it's impossible to make it work without changes to `rustc_driver`. I'll hopefully take a look at it soon and try to come up with a better overall solution. It should be possible to e.g. capture more data into the boxed closure, and the use of a `static` is somewhat incidental.

Though the most direct solution is to offer a `static` that *contains* the report URL (and maybe additional e.g. footer), and which can be changed independently of the panic hook.

Comment (bjorn3) [reply compiler/rustc_driver/src/lib.rs:1163]: You could have something like https://github.com/bjorn3/rustc_codegen_cranelift/blob/d498e6d6971575714353f11aa4d3e63a9d2030b2/src/bin/cg_clif.rs#L20-L34 except that you call `panic::take_hook()` twice and use the result of the second time as default hook. The first time you take the hook set by rustc_driver and ignore it. The second time you take the default libstd hook.

Comment (eddyb) [reply compiler/rustc_driver/src/lib.rs:1163]: > If no custom hook is registered, the default hook will be returned.

Ah, this behavior documented for `std::panic::take_hook` was what I missed, thanks!

**Pull request merged**